File: agent/options_walls.py
"""Phase 3: options call/put walls -- max-open-interest strikes read as dealer-hedging
pressure zones (Smart Money confluence, STRATEGY.md section 6). Informational only:
never a hard gate, never changes size or rejects a setup -- it only adds or removes a
confluence hit near a proposed entry, same as an S/R zone or a Fib level.

Yahoo's openInterest field is settlement-computed and occasionally comes back 0 across
every strike on every ticker (observed even on SPY/QQQ, which always carry real OI in
reality) -- a known gap in the free feed, not a bug in this code. When that happens we
fall back to today's traded volume as a much weaker "activity wall" proxy and label it
as such (source="volume_fallback"); if both are empty we report unavailable rather than
fabricate a level.
"""
from __future__ import annotations
from dataclasses import dataclass
import datetime as dt


@dataclass
class OptionsWalls:
    source: str = "unavailable"  # "open_interest" | "volume_fallback" | "unavailable"
    expiry: str | None = None
    days_to_expiry: int | None = None
    call_wall: float | None = None
    call_wall_strength: int | None = None
    put_wall: float | None = None
    put_wall_strength: int | None = None
    put_call_ratio: float | None = None
    dist_to_call_wall_pct: float | None = None
    dist_to_put_wall_pct: float | None = None


def _pick_expiry(expirations: list[str], min_days: int = 7, max_days: int = 45) -> tuple[str | None, int | None]:
    """Skip 0-6 day expiries (thin, near-worthless OI/quotes); prefer the first one
    inside [min_days, max_days] (front-month, most liquid); fall back to the nearest
    expiry >= min_days if nothing lands in that window."""
    today = dt.date.today()
    candidates = []
    for e in expirations:
        try:
            d = dt.date.fromisoformat(e)
        except ValueError:
            continue
        days = (d - today).days
        if days >= min_days:
            candidates.append((e, days))
    if not candidates:
        return None, None
    in_window = [c for c in candidates if c[1] <= max_days]
    return (in_window[0] if in_window else candidates[0])


def read(ticker: str, price: float, band_pct: float = 25.0) -> OptionsWalls:
    """Call/put wall read for one ticker. `price`: current price, used to (a) bound the
    strike search to a realistic band so a deep-OTM outlier can't win on a thin book,
    and (b) compute distance-to-wall percentages."""
    import yfinance as yf

    out = OptionsWalls()
    try:
        tk = yf.Ticker(ticker)
        expirations = list(tk.options)
    except Exception:
        return out
    if not expirations:
        return out

    expiry, days = _pick_expiry(expirations)
    if expiry is None:
        return out

    try:
        chain = tk.option_chain(expiry)
    except Exception:
        return out
    calls, puts = chain.calls, chain.puts
    if calls.empty and puts.empty:
        return out

    out.expiry, out.days_to_expiry = expiry, days
    lo, hi = price * (1 - band_pct / 100), price * (1 + band_pct / 100)
    calls = calls[(calls["strike"] >= lo) & (calls["strike"] <= hi)]
    puts = puts[(puts["strike"] >= lo) & (puts["strike"] <= hi)]

    for col in ("openInterest", "volume"):
        call_total = calls[col].fillna(0).sum() if col in calls else 0
        put_total = puts[col].fillna(0).sum() if col in puts else 0
        if call_total <= 0 and put_total <= 0:
            continue
        out.source = "open_interest" if col == "openInterest" else "volume_fallback"
        if call_total > 0:
            row = calls.loc[calls[col].idxmax()]
            out.call_wall = float(row["strike"])
            out.call_wall_strength = int(row[col])
            out.dist_to_call_wall_pct = round((out.call_wall / price - 1) * 100, 2)
        if put_total > 0:
            row = puts.loc[puts[col].idxmax()]
            out.put_wall = float(row["strike"])
            out.put_wall_strength = int(row[col])
            out.dist_to_put_wall_pct = round((out.put_wall / price - 1) * 100, 2)
        if call_total > 0 and put_total > 0:
            out.put_call_ratio = round(float(put_total / call_total), 2)
        break  # openInterest worked (or we've already fallen back to volume) -- stop

    return out

File: web/lib/optionsWalls.ts
/** Port of agent/options_walls.py's MarketData.app path only (Node/Vercel has
 * no yfinance, so the yfinance-OI and volume-fallback tiers from the Python
 * version aren't available here) -- informational confluence input only,
 * never a hard gate, same as the daily pipeline. Reports source:"unavailable"
 * rather than fabricate a level when no token is set or the request fails. */

export interface OptionsWalls {
  source: "open_interest" | "unavailable";
  provider: "marketdata" | null;
  expiry: string | null;
  days_to_expiry: number | null;
  call_wall: number | null;
  call_wall_strength: number | null;
  put_wall: number | null;
  put_wall_strength: number | null;
  put_call_ratio: number | null;
  dist_to_call_wall_pct: number | null;
  dist_to_put_wall_pct: number | null;
}

const UNAVAILABLE: OptionsWalls = {
  source: "unavailable", provider: null, expiry: null, days_to_expiry: null,
  call_wall: null, call_wall_strength: null, put_wall: null, put_wall_strength: null,
  put_call_ratio: null, dist_to_call_wall_pct: null, dist_to_put_wall_pct: null,
};

function pickExpiry(expirations: string[], minDays = 7, maxDays = 45): [string | null, number | null] {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const candidates: [string, number][] = [];
  for (const e of expirations) {
    const d = new Date(e + "T00:00:00Z");
    if (isNaN(d.getTime())) continue;
    const days = Math.round((d.getTime() - today.getTime()) / 86400000);
    if (days >= minDays) candidates.push([e, days]);
  }
  if (candidates.length === 0) return [null, null];
  const inWindow = candidates.filter(([, d]) => d <= maxDays);
  return inWindow.length > 0 ? inWindow[0] : candidates[0];
}

export async function readOptionsWalls(ticker: string, price: number, token: string | null, bandPct = 25.0): Promise<OptionsWalls> {
  if (!token) return UNAVAILABLE;
  try {
    const headers = { Authorization: `Bearer ${token}` };
    const expRes = await fetch(`https://api.marketdata.app/v1/options/expirations/${ticker}/`, { headers });
    const expJson = await expRes.json();
    if (expJson?.s !== "ok" || !expJson?.expirations?.length) return UNAVAILABLE;

    const [expiry, days] = pickExpiry(expJson.expirations);
    if (!expiry) return UNAVAILABLE;

    const lo = price * (1 - bandPct / 100);
    const hi = price * (1 + bandPct / 100);
    const chainUrl = new URL(`https://api.marketdata.app/v1/options/chain/${ticker}/`);
    chainUrl.searchParams.set("expiration", expiry);
    chainUrl.searchParams.set("strike", `${lo.toFixed(2)}-${hi.toFixed(2)}`);
    const chainRes = await fetch(chainUrl.toString(), { headers });
    const chainJson = await chainRes.json();
    if (chainJson?.s !== "ok") return UNAVAILABLE;

    const strikes: number[] = chainJson.strike ?? [];
    const ois: number[] = chainJson.openInterest ?? [];
    const sides: string[] = chainJson.side ?? [];
    if (!(strikes.length && ois.length === strikes.length && sides.length === strikes.length)) return UNAVAILABLE;

    let callBest: [number, number] | null = null;
    let putBest: [number, number] | null = null;
    let callTotal = 0, putTotal = 0;
    for (let i = 0; i < strikes.length; i++) {
      const oi = ois[i] || 0;
      if (sides[i] === "call") {
        callTotal += oi;
        if (!callBest || oi > callBest[1]) callBest = [strikes[i], oi];
      } else if (sides[i] === "put") {
        putTotal += oi;
        if (!putBest || oi > putBest[1]) putBest = [strikes[i], oi];
      }
    }
    if (callTotal <= 0 && putTotal <= 0) return UNAVAILABLE;

    const out: OptionsWalls = { ...UNAVAILABLE, source: "open_interest", provider: "marketdata", expiry, days_to_expiry: days };
    if (callBest) {
      out.call_wall = callBest[0];
      out.call_wall_strength = callBest[1];
      out.dist_to_call_wall_pct = Math.round(((out.call_wall / price - 1) * 100) * 100) / 100;
    }
    if (putBest) {
      out.put_wall = putBest[0];
      out.put_wall_strength = putBest[1];
      out.dist_to_put_wall_pct = Math.round(((out.put_wall / price - 1) * 100) * 100) / 100;
    }
    if (callTotal > 0 && putTotal > 0) out.put_call_ratio = Math.round((putTotal / callTotal) * 100) / 100;
    return out;
  } catch {
    return UNAVAILABLE;
  }
}
